scripts/cluster_network.py

```diff
@@ -106,6 +106,148 @@ def weighting_for_country(df: pd.DataFrame, weights: pd.Series) -> pd.Series:
+                    per_country=False,
+                )
+            )
+        busmap = pd.concat(busmaps)
+        return busmap.reindex(n.buses.index)
+
+    shapes.index = shapes.index.astype(str)
+    shapes.index.name = "name"
+    points = gpd.points_from_xy(**buses[["x", "y"]], crs=4326)
+    coords = gpd.GeoDataFrame(geometry=points, index=buses.index)
+    column = cluster_names or "name"
+    busmap = gpd.sjoin(coords, shapes, how="left")[column].rename("busmap")
+
+    if busmap.isnull().any():
+        unassigned = coords[busmap.isnull()]
+        # Take a projection which properly handles distances for European areas.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=DeprecationWarning)
+            unassigned_converted = unassigned.to_crs(3035)
+        shapes_converted = shapes.to_crs(3035).set_index(column)
+        for i, row in unassigned_converted.iterrows():
+            dists = shapes_converted.distance(row.geometry)
+            busmap.at[i] = dists.idxmin()
+
+    # split by sub-network, match target pattern "<country><sub_network> <number>"
+    if cluster_names is None:
+        nth_sub_network = buses.sub_network.loc[busmap.index]
+        split = busmap.str.split()
+        busmap = split.str[0] + nth_sub_network.astype(str) + " " + split.str[1]
+
+    return busmap
+
+
+def copperplate_buses(n: pypsa.Network, ds: pd.Series):
+    """
+    Copperplate buses that belong to the same group.
+
+    Based on the input pandas series, buses are grouped together into market zone by
+    replacing existing connections between the buses with a new connection of infinite capacity.
+
+    Parameters
+    ----------
+    n : pypsa.Network
+    ds : pd.Series
+        Series with index of buses and values of market zones.
+    """
+
+    # Remove connections between buses in the same zone
+    for c in n.branch_components:
+        df = n.static(c)
+        bus0_zones = df.bus0.map(ds).values
+        bus1_zones = df.bus1.map(ds).values
+        to_remove = df.index[bus0_zones == bus1_zones]
+        if len(to_remove) > 0:
+            n.remove(c, to_remove)
+
+    # Create reverse mapping from zone to buses
+    zone_to_buses = ds.groupby(ds).apply(lambda x: set(x.index))
+    # Add new lines with infinite capacity within each zone
+    for zone, buses in zone_to_buses.items():
+        if len(buses) > 1:
+            # Create lines between first bus and all others
+            first_bus = list(buses)[0]
+            other_buses = list(buses)[1:]
+
+            for i, bus in enumerate(other_buses):
+                n.add(
+                    "Link",
+                    f"copper_{zone}_{i}",
+                    carrier="copper",
+                    bus0=first_bus,
+                    bus1=bus,
+                    p_nom=float("inf"),
+                    p_min_pu=-1,
+                    underwater_fraction=0.0,
+                    under_construction=0.0,
+                )
```

Comment:
Doesn't that create many unnecessary links? Given that they will never be binding, they could also generate more degeneracies. Why not map all buses to bidding zone buses for all components, a bit like in https://pypsa.readthedocs.io/en/latest/examples/scigrid-redispatch.html#Build-market-model-m-with-two-bidding-zones?

```py
for c in m.iterate_components(m.one_port_components):
    c.static.bus = c.static.bus.map(zones)

for c in m.iterate_components(m.branch_components):
    c.static.bus0 = c.static.bus0.map(zones)
    c.static.bus1 = c.static.bus1.map(zones)
    internal = c.static.bus0 == c.static.bus1
    m.remove(c.name, c.static.loc[internal].index)
```

(and remove all original buses)

This would also be, in principle, identical to the regular clustering approach.

Reply:
this unfortunately does not work well with the rest of the workflow. there are loads of functions that rely on a clear country assignment of components (power plants, energy totals, etc.). If we allowed for buses that don't represent a single country but multiple, most of these functions would fails due to missing reference. keeping the countries distinct and copperplating them (we want to make this optional) is the smoothest options as far as I can see.

Reply:
Ah, so DE_LU is the only cross-country bidding zone. There would be one non-binding infinite link between the two, but the rest of the buses in Germany are clustered to one electricity bus for Germany (and same for Luxemburg)? That's fine! I just wouldn't do it for all buses in Germany/Luxemburg.

Reply:
yes, exactly! it only does it for the electrical country buses (there is only one bus for DE and LU)